source/dofs/dof_handler_policy.cc

```diff
@@ -3205,97 +3205,67 @@ namespace internal
                 std::vector<types::global_dof_index> local_dof_indices (cell->get_fe().dofs_per_cell);
                 cell->get_dof_indices (local_dof_indices);
 
-                Assert (std::find (local_dof_indices.begin(),
-                local_dof_indices.end(),
-                numbers::invalid_dof_index)
-                ==
-                local_dof_indices.end(),
-                ExcInternalError());
+                const bool is_complete
+                = (std::find (local_dof_indices.begin(), local_dof_indices.end(), numbers::invalid_dof_index)
+                == local_dof_indices.end());
+                Assert (is_complete, ExcInternalError());
 #endif
-                return std::vector<types::global_dof_index>();
+                return boost::optional<std::vector<types::global_dof_index>>();
               }
           };
 
           auto unpack
             = [] (const typename DoFHandlerType::active_cell_iterator &cell,
-                  const std::vector<types::global_dof_index> &dof_indices) -> void
+                  const std::vector<types::global_dof_index> &received_dof_indices) -> void
           {
-            // if we get an empty array, then the sending processor has
-            // determined that all data has already been sent. that can
-            // only happen if the complete data set had previously
-            // been sent, and in that case, (i) the user flag must have been
-            // deleted, and (ii) the current set of DoF indices must already
-            // be complete
+            // this function should only be called on ghost cells, and
+            // on top of that, only on cells that have not been
+            // completed -- which we indicate via the user flag.
+            // check both
+            Assert (cell->is_ghost(), ExcInternalError());
+            Assert (cell->user_flag_set(), ExcInternalError());
```

Comment:
Could it happen that we receive the same cell from two processors in the same round and we remove the flag after we unpack the first message? Let's talk about this if you have a minute.